### PR TITLE
fix: Rework relation field kinds

### LINKED ISSUE #2619: Add `Self` field kind
Required for (and broken out of) https://github.com/sourcenetwork/defradb/issues/2493 in order to prevent the schema-DAG from becoming circular.

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #2619 #2493

## Description

Reworks relation field kinds so that relations are not defined by names (which are mutable).

This PR replaces the old `ObjectKind` and `ObjectArrayKind`s with `CollectionKind`, `SchemaKind` and `SelfKind`. `NamedKind` has also been added - it allows users to interact with the definitions using their names (like before), before the kind gets translated into one of the other new kinds before storage.

Because Schemas must form a DAG (because of their CIDs (SchemaVersionID and Root), Schemas defined within the same action that form circular relations (e.g. User=>Dog=>User) have their CIDs grouped into a set with the same base id, plus an index relative to their lexographical order (name) within the set suffixed onto their CIDs.  This change is not user visible, besides the `-[index]` suffix on their CIDs.

Quite a lot had to change to facilitate this, and only a handful of things have been broken out to separate commits. Most of the work is in commit `Rework relation field kinds` - when reviewing that commit, I suggest starting with `tests/integration/collection_description/updates/replace/name_one_many_test.go` to see the original bug, followed by a quick scan of `client/schema_field_description.go`, then a look at `tests/integration/schema/self_ref_test.go` before properly reviewing the changes.  The bulk of the self reference logic is in `internal/db/schema_id.go`.

Please give me a shout if you have any questions! I think this should be the last structural change to collection/schema definitions :)

### CODE REVIEW

**Review by fredcarle — APPROVED** [review]
This is great work Andy! I like the complex schema being treated as a single schema ID with parts. The code is nice. There are only a few places where naming caused a bit of confusion on my end and I gave a few suggestion to alleviate it. Feel free to merge once the conversations are resolved.

File: client/definitions.go [hunk 1]
```diff
@@ -195,3 +201,167 @@ func (f FieldDefinition) GetSecondaryRelationField(c CollectionDefinition) (Fiel
...
+		// no-op
+	}
+
+	return CollectionDefinition{}, false
+}
+
+// GetDefinitionUncached returns the definition that the given [FieldKind] points to, if it is found in the given store.
+//
+// If the related definition is not found, or an error occurs, default and false will be returned.
+func GetDefinitionUncached(
```
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply client/definitions.go:286]: I like this, thanks.

- [ ] Rename GetDefinitionUncached

File: internal/db/definition_validation.go [hunk 2]
```diff
@@ -30,44 +30,62 @@ type definitionState struct {
...
+		schemaByName:      schemaByName,
+		definitionsByName: definitionsByName,
+		definitionCache:   client.NewDefinitionCache(definitions),
+	}
+}
 
-		if _, ok := schemaVersionsAdded[schema.VersionID]; ok {
-			continue
-		}
+// newDefinitionState creates a new definitionState object given the provided
```
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply internal/db/definition_validation.go:70]: Not a problem, I'll have a look and tweak them a bit.

-  [ ] Review `newDefinitionState`s plus docs

File: internal/db/definition_validation.go [hunk 3]
```diff
@@ -843,6 +845,58 @@ func validateFieldNotDuplicated(
 	return nil
 }
 
+func validateSelfReferences(
+	ctx context.Context,
+	db *db,
+	newState *definitionState,
+	oldState *definitionState,
```
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply internal/db/definition_validation.go:852]: I'm probably going to leave this one as is, this func (and most other funcs in this file) satisfies an interface and they all have the same prop names and I think I prefer consistency over worrying about which funcs use which props (also aids copy-pasting the sig).

- [ ] Have a quick look when looking at `newDefinitionState`

File: internal/db/schema_id.go [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+
+	schemaSets := getSchemaSets(newSchemaPtrs)
+
+	for _, schemaSet := range schemaSets {
+		setID, err := generateSetID(schemaSet)
+		if err != nil {
+			return err
+		}
+
+		setIDs(setID, schemaSet)
```
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply internal/db/schema_id.go:45]: Renamed to `assignIDs`, as the func does not generate the ids

File: internal/db/definition_validation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply internal/db/definition_validation.go:70]: Docs tweaked, renamed the first col-only func to `newDefinitionStateFromCols` (`newDefinitionState` should be the default func to use)

File: internal/db/definition_validation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndrewSisley) [reply internal/db/definition_validation.go:852]: Leaving as-is

File: internal/db/definition_validation.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fredcarle) [reply internal/db/definition_validation.go:70]: `FromCols` is not really clarifying anything though as they are both using collectionSomething.

**Pull request merged**